### DISCUSSION TITLE
[Proposal] File-scoped types in explicit interface implementations.

### DISCUSSION THREAD

**Discussion by ShadedBlink** [dotnet/csharplang#8357] [Language Ideas]
Allow file-scoped types in explicit interface implementations. Those members are inaccessible without explicit conversion to given interface, thus it is safe to allow them use file-scoped types in their declarations.
```csharp
public class PublicType
    // This is already supported since C# 11.
    : IList<FileScopedType> // Totally okay, we can declare interfaces with file-scoped type since no one can cast our `PublicType` instance to `IList<FileScopedType>` outside of this file.
{
    // This is still restricted in C# 12.
    void IList<FileScopedType>.Add(FileScopedType arg) // Totally okay, since no one can access this member without casting this type to `IList<FileScopedType>`, what is obviously possible only in current file.
    {
        throw null;
    }
}

file class FileScopedType {
}
```

**Comment by TahirAhmadov** [comment]
How is this different from other non-public access levels, like `internal`?
Either way, it should be like this:
```cs
public class PublicType
    // This is already supported since C# 11.
    : explicit IList<FileScopedType> // has to have "explicit" keyword (don't know if it should be before or after)
{
    void IList<FileScopedType>.Add(FileScopedType arg)
    {
        throw null;
    }
    public void Add(FileScopedType arg) { } // compilation error: inconsistent accessibility, and 
    // if explicit method above is absent, also "method Add not implemented"
}

file class FileScopedType {
}
```

**Reply by ShadedBlink** [reply]
It is different in the way how it is treated in current assembly. Sometimes you want to isolate some logic inside the assembly and don't want to extract it to a new assembly. Thus you can isolate such logic in given file, and it won't interfere with intellisense and code in other classes. For example, you have multiple classes that use some individual helper classes/structures with similar names, but they share no domain logic. Of course, we can just declare them internal and isolate in own namespace, but they will appear in intellisense suggestions anyway. Also we can make them private, but what is the point of file-scoped types then?
My main idea is that we already have private/internal interface implementations. We can use them only in current assembly or in current class. I want to extend file-scoped logic to allow interface usage only in current file(I.e. more than a class visibility, but less than an assembly visibility).

**Comment by jaredpar** [comment]
>  Totally okay, since no one can access this member without casting this type to `IList<FileScopedType>`, what is obviously possible only in current file.

That is not correct. It can be accessed with generics. Consider the following: 

```csharp
M(new PublicType());
void M<T>(IList<T> list) { 
  list.Add(null);
}
```

In this case `T` would be `FileScopedType`.  That by itself isn't really violating the principal of file scoped types because you can't really do anything with `T`. But it's easy to extend the samples a bit where you can start probing into the file scoped type. Consider the following: 

```csharp
// file1 
class Col: IEnumerable<FileType> {
    public IEnumerator<FileType> GetEnumerator() => throw null;
    IEnumerator IEnumerable.GetEnumerator() => throw null;
}

class FileType : IDisposable 
{
    public void Dispose() => throw null;
}

// file2 
var col = new Col();
M(col);
void M<T>(IEnumerable<T> e) where T : IDisposable
{
  foreach (var t in e) {
    t.Dispose(); // Calling File scoped type member from diff file
  }
}
```

The ability to _observe_ the shape of a file scoped type like this goes very much against their design.

**Reply by ShadedBlink** [reply]
Still such probing is equal to dependency on actual internal implementation of the type which is bad anyway. But we can do the same thing, using private type instead of file-local type. In such case we get `CS0122`, thus we can use implementation only in current class and not outside, what is pretty close to what I propose.
Thus we can change this proposal's logic so those file local types are not visible in other assemblies(already implemented), and attempt to cast `PublicType` to generic types/interfaces that are declared using `FileLocalType` will result in something close to `CS0122` because `FileLocalType` is inaccessible outside of it's declaration's file.
I.e. in your case your call to `M(col)` will work in file1, but fail to compile with error `CS0122`(or similar) in file2.

**Reply by jaredpar** [reply]
> Still such probing is equal to dependency on actual internal implementation of the type which is bad anyway

Yes but file types were specifically designed to prevent this type of probing. The original motivation is that source generator authors wanted to be able to freely evolve their implementation without fear that customers were taking a dependency on them. The moment a file type can be accessed in this manner it cuts into that motivation a bit.

> But we can do the same thing, using private type instead of file-local type. In such case we get CS0122, thus we can use implementation only in current class and not outside, what is pretty close to what I propose.

Even errors like this produce observable behavior. For example, it can be the reason a candidate is filtered out of overload resolution leaving a single candidate that is accepted. A file type evolves in the future to no longer implement an interface and now it changes how that overload resolution works in a potentially observable way.

The premise of file types is that developers can use them with confidence that they're not observable outside the current file. This is very important for customers like source generator authors who want to be able to freely evolve their implementation over time. Design changes to file local types need to keep that firmly in mind because designs that break that principal are likely to be rejected. It's hard to see a way to have accessible interface implementations use file local type parameters without being observable.

**Reply by ShadedBlink** [reply]
> Even errors like this produce observable behavior. For example, it can be the reason a candidate is filtered out of overload resolution leaving a single candidate that is accepted. A file type evolves in the future to no longer implement an interface and now it changes how that overload resolution works in a potentially observable way.

I think I am missing the point there. For example, we have this class:
```csharp
public class PublicType : IEnumerable<PublicType.PrivateType> {
private struct PrivateType {}
}
```
Developers introduce `PrivateType` as private because they want to make sure that nobody will be able to use it since it is an implementation detail. Because of that nobody will be able to perform any type of dependency on the fact that this class implements `IEnumerable<T>` or `IEnumerable<PrivateType>` due to `CS0122`.
If you want to make the fact of this interface implementation completely hidden from outside world, wouldn't it make more sense to just correct overload seeking algorithm? Like exclude interfaces with inaccessible types and instead show error like `PublicType doesn't implement interface IEnumerable<T> or generic parameter is inaccessible`. Because actually what we have is that private/internal types are used for the very same purpose - they are internal implementation details and no one should ever know about them. They should be unobservable outside of their visibility scope, that's exactly the reason of existence of accessibility modifiers.
I believe that `file` is just another level of visibility between `internal` and `private` and thus it should follow common rules of accessibility modifiers. If you want to ensure unobservability of file local types, then use same rules for private and internal types.

**Reply by jaredpar** [reply]
> Because of that nobody will be able to perform any type of dependency on the fact that this class implements IEnumerable<T> or IEnumerable<PrivateType> due to CS0122.

The issue though is CS0122 is _observable_. Any time you have a conversion error that impacts operations like overload resolution, lambda inference, etc ... This error can be necessary for those operations to succeed and hence the presence or absence of this interface implementation can change whether the program compiles or not. 

> ... wouldn't it make more sense to just correct overload seeking algorithm

That would be a breaking change in itself at this point. I'm also guessing that it would have other interesting impacts as interface behavior can be very tricky in corner areas like this

> I believe that file is just another level of visibility between internal and private and thus it should follow common rules of accessibility modifiers.

LDM considered deeply the idea of whether or not `file` should an accessibility modifier and decided against that interpretation. Instead it is a type modifier that is separate from accessibility.

**Reply by ShadedBlink** [reply]
> The issue though is CS0122 is observable

Okay, lets use `CS0311` then. Nobody can see this interface outside, thus, it's expected that in their eyes this class doesn't implement given interface. Outside of the file you just see an error that given type doesn't implement such interface at all - same case when it is really not implemented and when it is file-local implementation.

> That would be a breaking change in itself at this point.

It can be a breaking change only if someone already used such dependency on file-local interface implementation in their code outside of given file, what I expect to be highly unlikely.
However, there were already precedents when we restricted already possible cases with file-local types. I mean the case when it was possible to create an internal descendant of file-local type. It was restricted later in `net8.0`. I see this proposal in the same way - it wasn't expected to work this way, thus, it would be right to stop it now before it goes any further.
What's for overload seeking algorithm, it should not be a breaking change in any other case, since we just disable file-local implementations outside of file scope. I am open for your ideas on how anybody else could be affected by such change because I don't see any changes to their behaviour.

**Reply by jaredpar** [reply]
> It can be a breaking change only if someone already used such dependency on file-local interface implementation in their code outside of given file, what I expect to be highly unlikely.

This would require corrections to the general overload resolution algorithm, not specific to `file` local types. This is why I classify it as a breaking change.

**Reply by ShadedBlink** [reply]
@jaredpar , so according to your answer you classify this proposal as breaking change not because of behaviour changes, but because of the fact that you will have to edit overload resolution algorithm and afraid to break anything along the way. Have I understood you right?
Do you see any notable issue with behaviour changes if we assume that you will implement this proposal without breaking anything?

**Reply by jaredpar** [reply]
> Do you see any notable issue with behaviour changes if we assume that you will implement this proposal without breaking anything?

I'm very skeptical this proposal can be reasonably implemented while maintaining the invariant that `file` types are not observable outside the file they are defined in. Interface implementation details can show up in a lot of different ways. The overload case I've brought up is just one of them.

**Reply by ShadedBlink** [reply]
Well, that's exactly what I want to get from you. I want to get a list of problems for this proposal so I can find a resolve for them.
The main problem that I see is that file-local interfaces are **NOT** forbidden at the moment. I.e. my proposal is **PARTIALLY** implemented and **PARTIALLY** violating your concerns. The only interfaces that can be implemented in such way are interfaces that doesn't have members with generic argument mentioned, i.e. if there will be an empty generic interface, we are free to imlement it with file-local type, which is obviously what you want to avoid. In such case we have to restrict this ability as well.
However, it limits file-local types to be just a context-objects which can live only inside a method scope. There is definitely a need to have an utility-types that are limited to current file. For example, some helper classes that doesn't have any use outside of this file. Of course, we can make them private, but same goes for the current file-local types - we can make them private.
I am okay with just total disable of file-local interface implementations, but we need something in exchange - `file` as access modifier, that just limits type resolution scope to given file without any _"super privacy"_. Maybe make it `file internal` in opposite to current `file private` modifier.

**Reply by ShadedBlink** [reply]
IMHO, it would be right to make current `file` keyword a true accessibility modifier(between `private` and `internal`), it won't change any existing compilation code. People will just get a way to use it in more ways than before. Then "supper privacy" case should be a `file private` modifier or just a `private` outside of type, i.e. when private is used outside of type, then it is private to file(as `file` is the next outer scope).
True file-local types(as accessibility modifiers) is a too powerfull tool for clean-code since it allows us to disassemble current mountains of nested private types to just file local registrations. It is such a waste to just decline it in a favour of these method-scoped types.

**Reply by ShadedBlink** [reply]
@jaredpar , got your response on mail, but don't see it there.

>Again, LDM considered this at length and decided they explicitly did not want this behavior and preferred that file be a very different feature than an accessibility modifier.

I am okay with having such `file` "super privacy" case, I just want to say that accessibility modifier case would be very usable as well. I would say we need both of them, and I am sure that later case will be much more popular.

For your example, there is a **compilable** code. It works right now and there is no restriction for it in `net8.0`. We are allowed to make file-local implementations unless they contain any member with file-local type in signature. So your observability protection doesn't work at the moment.
```csharp
// File1.cs
void M<T>(IIsRelatedTo<T> i1) {
Console.WriteLine($"{i1.GetType()} is related to type {typeof(T)}");
}

// File2.cs
public class C1 : IIsRelatedTo<C2> {
void IIsRelatedTo<C2>.DoSomething() => throw null!; // no error!
}
public interface IIsRelatedTo<T> {
void DoSomething();
}
file class C2 { }
```

> To do so you'd probably need to say that the implementation of I1<C2> is only observable on C1 when you are inside File2.cs. That would solve the problem but it would also be a fairly significant change to the language. It would mean for example you can't ask what interfaces a type implements, you have to say from file1.cs what interfaces does the type implement.
The other way to fix this is to loosen the invariant around file local types to say it's okay if they're observed in this specific way.

I understand that file-involved discoverability is not an easy thing to do, thus implementing a new accessibility modifier case of `file` with standard `CS0122` should be an option. I wonder why you think that we should leave exactly one way to handle file-visibility when we can support both of them. It's not like support of one of them makes impossible support of another one.

**Reply by ShadedBlink** [reply]
By the way, I suddenly found a way to overcome this problem. It is compilable, it is working, but it is awfull. 
```
public class C1 : IIsRelatedTo<C2> {
	void IIsRelatedTo<C2>.DoSomething(ref byte arg) => throw null!; // no error!

	void TestMethod() {
		var x1 = new C1();
		var x2 = new C2();
		x1.DoSomething(ref x2); // no error!
	}
}
public interface IIsRelatedTo<T> {
	void DoSomething(ref byte arg);
}
file class C2 { }
internal static class Extensions {
	public static void DoSomething<T>(this IIsRelatedTo<T> origin, ref T arg)
			=> origin.DoSomething(ref Unsafe.As<T, byte>(ref arg));
}
```

**Reply by jaredpar** [reply]
>  got your response on mail, but don't see it there.

Cause I decided I wanted to rethink one part of it so I deleted the message.

**Reply by jaredpar** [reply]
> Well, that's exactly what I want to get from you. I want to get a list of problems for this proposal so I can find a resolve for them.

A significant part of feature design is discovering the problems. It's a task that often takes a significant time commitment. This isn't a case where I know the problems and are refusing to divulge them, instead it's a case of it's a tricky area and no one has sat down and paged it all in to think through what the problems are. Part of proposing a feature is doing that work. 

Consider as an example the [proposal I filed](https://github.com/dotnet/csharplang/pull/8413) recently for allowing `yield` in a `try / catch`. That is a relatively simple change, just remove a restriction around `yield`. But it took me 3-4 hours of digging into the topic, chatting with people involved in related features or the original design to fully understand the set of problems that I needed to solve. Anything with interfaces and visibility is going to be significantly more complex than that. 

> So your observability protection doesn't work at the moment.

Correct. This is one of the reasons I deleted my answer last night. I thought through a couple of cases where I could expose the type today that were similar. The guidance to `file` authors is if you have that in a impls / base list then there is some amount of external exposure that you're taking on. 

The explicit interface proposal does up the exposure to a new level because it allows you to get instances of the types via a type parameter. For example: 

```csharp
// file1.cs
M(new C1());
void M<T>(I1<T> p)
{
    // T is a file type now
    T local = p.Prop;
}

// file2.cs
class C1 : I1<C2>
{
    C2 I1<C2>.Prop => new C2();
}
interface I1<T>
{
    T Prop { get; }
}
file class C2 { }
```

Essentially now I can get `T` instances which are `file` types in another file. That's an escalation of capabilities, not sure how severe it is / isn't. 

Other areas I would want to dig down into is:

- How does this impact binary compat? Every change you make to the file will change the type identity of a `file` type within it. Now that these can be exposed through interfaces are we creating any subtle binary compat issues. Basically is there a way for code outside this assembly to now depend on `file` type since it was exposed via an interface? If so that would be pretty fragile. 
- Can I use `partial` to get to the file type? Essentially if the class implementing the interface is split between files via `partial` can I use that to get access to the `file` type outside the file? Basically split the interface implementation up amongst the partial declarations. I _think_ that is okay because explicit implementation requires the type argument to be stated but would want to dig into all those rules. 

>  thus implementing a new accessibility modifier case of file with standard CS0122 should be an option. I wonder why you think that we should leave exactly one way to handle file-visibility when we can support both of them

If you want to push the issue of `file` being an accessibility modifier I think you should file a separate proposal that is specifically aimed at it. As I've mentioned a few times, C# LDM (not me specifically) considered and rejected this idea. Suggest looking through the design meetings, internalizing the rationale and providing reasons why you think the decision was incorrect.

**Reply by ShadedBlink** [reply]
> This isn't a case where I know the problems and are refusing to divulge them

I am sorry if you had a feeling that I sense any offense from you or etc. I am just curious about implementation as well too and want to dive deeper in subject because I don't know everything as well. Thus I am probing in problems and proposal variations.

> The explicit interface proposal does up the exposure to a new level because it allows you to get instances of the types via a type parameter.

Actually I don't think that given case is the problem as for now we are free to store file-scoped types in object fields of a public class. Even more, if file-scoped type implements any interface or base type, we are free to publicly store it as given type. It was my previous implementation before refactoring, so I guarantee that following implementation is working.
```csharp
class C1
{
    public I1 Field = new C2();
}
interface I1
{
}
file class C2 : I1 { }
```

Thus we get in the situation when we are free to display our file-scoped objects publicly in any way except when it is used as generic argument of interface. No offense, but that sounds totally stupid.

>How does this impact binary compat? Every change you make to the file will change the type identity of a file type within it. Now that these can be exposed through interfaces are we creating any subtle binary compat issues. Basically is there a way for code outside this assembly to now depend on file type since it was exposed via an interface? If so that would be pretty fragile.

As we already know, we have no way to make direct dependency on file-scoped type due to lack of type discoverability. Thus recompilation of library with new name of such type definitely won't break anything. So we have only indirect internal dependency left, i.e. user code should depend on the fact that object's type implements any interface or has some other property like "`Type.Name` starts with `FileLocalType`". This is absolutely same as dependency on internal implementation, so responsibility is on the user code. If we allow users to get instances of file-scoped types(as in previous example), I see no problem with them getting instances of those types by generic arguments.

> Can I use partial to get to the file type? Essentially if the class implementing the interface is split between files via partial can I use that to get access to the file type outside the file? Basically split the interface implementation up amongst the partial declarations. I think that is okay because explicit implementation requires the type argument to be stated but would want to dig into all those rules.

`partial` definitely should **not** allow us to get file-scoped types. `file` limits to file, not to type like `private`. More so, this modifier should be especially usefull when you split complex type in different files. Thus we can leave some logic exclusive to one part of the class, other part is exclusive to other part. If we extend `file` to `partial`, then such split is impossible.

> If you want to push the issue of file being an accessibility modifier I think you should file a separate proposal that is specifically aimed at it. As I've mentioned a few times, C# LDM (not me specifically) considered and rejected this idea. Suggest looking through the design meetings, internalizing the rationale and providing reasons why you think the decision was incorrect.

It would really be much easier to just create another proposal, but current implementation of privacy of `file` just can't leave me be. It doesn't hide file-scoped objects at all, only the type, i.e. it works exactly like an accessibility modifier. The only difference is that we can't create file-scoped interfaces. Just try to look at other difference for this case from `public`-`private` relation - you won't see any. This protection you told about is implemented so bad that I even had no idea that it was a target.

- You can't create `public` members with `private` types - same for `file`.
- You can create `public` members with `private` objects by their parent or interface type - same for `file`.
- You can't inherit `private` base type with `public` descendant type - same for `file`.
- You can't inherit `public` base generic type with `private` generic argument - same for `file`.
- You can implement `private` interface in `public` type - same for `file`.
- You can implement `public` generic interface with `private` type - same for `file`.
- You can implement `public` generic interface with `private` generic argument - same for `file`.
- You can't explicitly implement members of `public` generic interface with `private` generic argument in declarations - same for `file`.
- You can implicitly implement members of `public` generic interface with `private` generic argument in declarations - same **is not possible** for `file`.

**Reply by ShadedBlink** [reply]
@jaredpar , any thoughts on this one for all this time? As of C#13, the only difference in behaviour between `file` and normal accessibility modifier is still only in generic interface implementation, everything else is similar.

**Reply by CyrusNajmabadi** [reply]
It's basically asking for a lot of work (especially investigation wise) for marginal gain. Could you consider not using a file-type in this case?

**Reply by ShadedBlink** [reply]
@CyrusNajmabadi , of course I can not use it, I can even live without a `for` clause and replace it with `if...goto`, but I believe it to be not a good idea.
I already provided here analysis on differences between `file` and `private`, which heavily contradict with @jaredpar explanation that `file` has a different purpose from a normal accessibility modifier, and is meant to keep objects(not types) isolated from outer world. `file` doesn't secure objects at all, it only interferes with type resolving, as it is meant for accessibility modifiers. I am okay if there will be atleast an issue created noting this proposal, even if nobody will take it in for development for a long time. As it stated in issue template, it should be invited by team-member, so I need somebody to approve this to create an issue.

**Reply by ShadedBlink** [reply]
What I mean is that I see that given generic interface implementation restriction is wrong and it should be lifted. I just want to either get this vision accepted as an issue or get some explanation for why it should be left as is. I am okay if it is not implemented for a long time due to implementation complexity since dotnet devs will be atleast aware of this bug. At the moment I still have not got any valid explanation for refusal.

**Reply by CyrusNajmabadi** [reply]
> At the moment I still have not got any valid explanation for refusal.

Jared's reasons seem valid to me. Regardless,  if a team member wants to champion this, they can do so.

**Reply by jaredpar** [reply]
> I already provided here analysis on differences between file and private, which heavily contradict with @jaredpar explanation that file has a different purpose from a normal accessibility modifier, and is meant to keep objects(not types) isolated from outer world.

To be clear: it is the explicit view of LDM that this is not an access modifier. There is overlap in functionality but it is meant to serve a different purpose than traditional access modifiers. This is not a case of my pushing my opinion on the behavior here (I was actually for having it be an accessibility modifier but we decided the other direction). 

> I am okay if there will be atleast an issue created noting this proposal, even if nobody will take it in for development for a long time

Issues are [only created](https://github.com/dotnet/csharplang/blob/main/Design-Process.md#proposed-feature) now when there is an LDM member willing to sponsor the proposal. It is not created just to note an idea, that is what discussions are for now. 

> At the moment I still have not got any valid explanation for refusal.

There are some issues for which there is an explicit "no we can't do this because X". This issue falls more into the category of "probably works" but also needs deeper vetting and has to cross the "is it worth the -100 points all proposals start with bar". It's not that we're refusing to do this, it's just that none of the LDM members are thus far inspired to champion this.

**Reply by ShadedBlink** [reply]
> Issues are [only created](https://github.com/dotnet/csharplang/blob/main/Design-Process.md#proposed-feature) now when there is an LDM member willing to sponsor the proposal. It is not created just to note an idea, that is what discussions are for now.

Well, that's enough for me. Then I will just hope that sometime later this will be finally fixed.

**Reply by martin-steinhoefer** [reply]
> > The issue though is CS0122 is observable
> 
> Okay, lets use `CS0311` then. Nobody can see this interface outside, ...

That is obviously **not** true. The list of implemented interfaces is part of the declaration of `PublicType` and as such is public knowledge as the type itself is public. That can be proven like this
```cs
var isEnumerable = (new PublicType() is IEnumerable); // will return true
```
although `IEnumerable` is not `IEnumerable<T>`, that line shows, that **any** interface implemented by a public type is **visible**, even if one can't write the name of that interface down in C# (you can't write down `IEnumerable<PublicType.PrivateType>`, b/o PrivateType is inaccessible). However, via reflection one is able to get the inferred type object of the generic type argument for further use.
